### PR TITLE
add lwt.2.5.2

### PR DESCRIPTION
**Pull request by pirbo**
This not ready, at least support for ppx has to be added.
I open a pull request to receive comments/advices.

### CODE REVIEW

File: packages/lwt-windows.2.5.2/opam [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+bug-reports: "http://github.com/ocsigen/lwt/issues"
+license: "LGPL with OpenSSL linking exception"
+
+build: [
+  ["ocaml" "setup.ml" "-configure"
+    "--override" "os_type" "Win32"
+    "--override" "flexlink" "%{lib}%/flexdll-windows/flexlink"
+    "--override" "ext_dll" ".dll"
+    "--override" "datarootdir" "%{prefix}%/windows-sysroot/share"
+    "--disable-pthread"
```
Comment (whitequark) [inline packages/lwt-windows.2.5.2/opam]: This should be enabled.

File: packages/lwt-windows.2.5.2/opam [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    "--disable-ppx"]
+  ["ocamlbuild" "-use-ocamlfind" "-just-plugin"]
+  ["env" "OCAMLFIND_TOOLCHAIN=windows" "ocaml" "setup.ml" "-build"]
+]
+install: [["env" "OCAMLFIND_TOOLCHAIN=windows" "ocaml" "setup.ml" "-install"]]
+remove: [["ocamlfind" "-toolchain" "windows" "remove" "lwt"]]
+depends: [
+  "ocaml-windows"
+  "ocamlfind" {build & >= "1.5.0"}
+  "base-bytes"
```
Comment (whitequark) [inline packages/lwt-windows.2.5.2/opam]: Not needed, there's only 4.02+ cross-compilers.

File: packages/lwt-windows.2.5.2/opam [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  "ocamlbuild" {build}
+#  "ppx_tools"
+]
+depopts: [
+  "base-threads"
+  "base-unix"
```
Comment (whitequark) [inline packages/lwt-windows.2.5.2/opam]: Not needed.

File: packages/lwt-windows.2.5.2/opam [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+remove: [["ocamlfind" "-toolchain" "windows" "remove" "lwt"]]
+depends: [
+  "ocaml-windows"
+  "ocamlfind" {build & >= "1.5.0"}
+  "base-bytes"
+  "ocamlbuild" {build}
+#  "ppx_tools"
+]
+depopts: [
+  "base-threads"
```
Comment (whitequark) [inline packages/lwt-windows.2.5.2/opam]: Not needed.

**Comment by whitequark** [comment]
Other than that and the ppx_tools issue looks fine.

File: packages/lwt-windows.2.5.2/opam (hunk 1 above)
Comment (pirbo) [inline packages/lwt-windows.2.5.2/opam]: It is copy/paste from https://github.com/vouillon/opam-windows-repository/blob/master/packages/win64-lwt/win64-lwt.2.4.5/files/build.sh 
Naive compilation with --enable-pthread fails with
```
/home/pirbo/.opam/4.02.3/bin/ocamlfind ocamlc -g -ccopt -I/usr/include -ccopt -Isrc/unix -c src/unix/lwt_unix_stubs.c
+ /home/pirbo/.opam/4.02.3/bin/ocamlfind ocamlc -g -ccopt -I/usr/include -ccopt -Isrc/unix -c src/unix/lwt_unix_stubs.c
In file included from /usr/include/stdlib.h:24:0,
                 from /home/pirbo/.opam/4.02.3/windows-sysroot/lib/ocaml/caml/misc.h:27,
                 from /home/pirbo/.opam/4.02.3/windows-sysroot/lib/ocaml/caml/alloc.h:21,
                 from src/unix/lwt_unix_stubs.c:28:
/usr/include/features.h:385:23: fatal error: gnu/stubs.h: Aucun fichier ou dossier de ce type
compilation terminated.
Command exited with code 2.
```

Comment (whitequark) [inline packages/lwt-windows.2.5.2/opam]: `lwt_unix_stubs.c`. You need to figure out what condition makes lwt select between unix and win32 stubs.

**Comment by pirbo** [comment]
My last commit may be improved but it allows to compile (and run with wine64)
```
let main () =
  Lwt_io.write_line Lwt_io.stdout "foo" >>
  Lwt_io.write_line Lwt_io.stdout "bar"

let%lwt () = main ()
```
using command line ```OCAMLFIND_TOOLCHAIN=windows ocamlbuild -use-ocamlfind -pkg lwt.unix -pkg lwt.ppx test.native```

File: packages/lwt-windows.2.5.2/opam (hunk 1 above)
Comment (pirbo) [inline packages/lwt-windows.2.5.2/opam]: It looks like lwt never use pthread under windows.
At least, if you look at its continuous integration test under cygwin: https://ci.appveyor.com/project/aantron/lwt
you see l.372 ```Use pthread: ......................................... false```

Comment (whitequark) [inline packages/lwt-windows.2.5.2/opam]: Cygwin is completely irrelevant to opam-cross-windows, this is a native build (using mingw). Cygwin is an Unix environment.

Comment (pirbo) [inline packages/lwt-windows.2.5.2/opam]: the _oasis file of lwt then:
```
Flag pthread
  Description: Use pthread
  Default$: !os_type(Win32)
```
I'm sorry but I won't try to enable-pthread myself.
(which does not mean I'm mad, I'm extremely thankful for your patience and all your advices)

Comment (whitequark) [inline packages/lwt-windows.2.5.2/opam]: Hm, maybe it isn't available on win32. Nevermind then.

File: packages/lwt-windows.2.5.2/opam [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    "--override" "ext_dll" ".dll"
+    "--override" "datarootdir" "%{prefix}%/windows-sysroot/share"
+    "--disable-pthread"
+    "--prefix" "%{prefix}%/windows-sysroot"
+    "--disable-libev"
+#    "--%{camlp4:enable}%-camlp4"
+    "--disable-camlp4"
+    "--%{react-windows:enable}%-react"
+    "--%{ssl-windows:enable}%-ssl"
+    "--%{base-unix:enable}%-unix"
```
Comment (whitequark) [inline packages/lwt-windows.2.5.2/opam]: This should just be always enabled, for the same reason we don't have a dependency or a depopt.

File: packages/lwt-windows.2.5.2/opam [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    "--%{base-threads:enable}%-preemptive"
```
Comment (whitequark) [inline packages/lwt-windows.2.5.2/opam]: Ditto.

File: packages/lwt-windows.2.5.2/opam [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    "--disable-ppx"]
+  ["ocamlbuild" "-use-ocamlfind" "-just-plugin"]
+  ["env" "OCAMLFIND_TOOLCHAIN=windows" "ocaml" "setup.ml" "-build"]
+]
+install: [
+  ["env" "OCAMLFIND_TOOLCHAIN=windows" "ocaml" "setup.ml" "-install"]
+  ["ln" "-s" "../../bin/ppx_lwt" "%{prefix}%/windows-sysroot/bin/ppx_lwt"]
+
+  # tell ocamlfind that package lwt.ppx is 'available'
+  ["ln" "-s" "../../../lib/lwt/ppx.cma"
```
Comment (whitequark) [inline packages/lwt-windows.2.5.2/opam]: Ok, this seems really gross. Can you please use sed to remove the `enable_if ...` line? It should be something like:
```
["sed" "-e" "/enable_if .../d" "-i" "%{prefix}%/windows-sysroot/lib/lwt/META"]
```

File: packages/lwt-windows.2.5.2/opam [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    "%{prefix}%/windows-sysroot/lib/lwt/ppx.cma"]
+  ["ln" "-s" "../../../lib/lwt/ppx_lwt.cmi"
+    "%{prefix}%/windows-sysroot/lib/lwt/ppx_lwt.cmi"]
+  ["ocamlfind" "-toolchain" "windows"
+    "ocamlopt" "-a" "-o" "%{prefix}%/windows-sysroot/lib/lwt/ppx.cmxa"]
+]
+remove: [
+  ["rm" "-f" "%{prefix}%/windows-sysroot/bin/ppx_lwt"]
+  ["rm" "-f" "%{prefix}%/windows-sysroot/lib/lwt/ppx*.cm*"]
```
Comment (whitequark) [inline packages/lwt-windows.2.5.2/opam]: This isn't necessary even without the `sed` hack since that's what `ocamlfind remove` does.

**Comment by whitequark** [comment]
Alright--once the last round of comments is addressed I'll merge it!

**Comment by pirbo** [comment]
0/ it is good to be demanding (see 2/)
1/ Sorry for the delay
2/ Sorry I miss a typo in react-windows opam file! I thought I tested my stuff ... Anyway I include the fix here (in a different commit) if it is OK
3/  The sed command is more general than what you suggested: I use ```sed -e '/"ppx.cm/p'``` that removes
```
 archive(byte) = "ppx.cma"
 archive(byte, plugin) = "ppx.cma"
 archive(native) = "ppx.cmxa"
 archive(native, plugin) = "ppx.cmxs"
 exists_if = "ppx.cma"
```
If I keep the archive lines, the linking fails because ocamlopt looks for ppx.cmxa that does not exist.

**Comment by whitequark** [comment]
Merged (I cherry-picked with minor stylistic change).

**Pull request closed without merging**